### PR TITLE
Add eng/common break check

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Extend `eng/common/scripts/Invoke-GitHubAPI.ps1:Get-GitHubPullRequest` to return files in a given pull request.
Add pipeline to prevent changes to `eng/common` that don't follow the workflow from being merged in

### CODE REVIEW

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 1]
```diff
@@ -98,10 +98,15 @@ function Get-GitHubPullRequest {
     $PullRequestNumber,
     [ValidateNotNullOrEmpty()]
     [Parameter(Mandatory = $true)]
-    $AuthToken
+    $AuthToken,
+    [Boolean]$ReturnFiles=$false #Pass true to return the files in the PR
```
Comment (benbp) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:102]: Would this be better as a `[Switch]`?

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 2]
```diff
@@ -98,10 +98,15 @@ function Get-GitHubPullRequest {
...
   )
 
   $uri = "$GithubAPIBaseURI/$RepoOwner/$RepoName/pulls/$PullRequestNumber"
+  if ($ReturnFiles) 
+  {
+    $uri += "/files"
```
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:108]: This change looks unrelated to the other change. What is the purpose of this change?

File: eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+trigger: none
+
+pr:
```
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml:3]: This looks like you are planning to run this as its own job. I'm not sure I'd go that far. I'd suggest instead we add it as a template step and include it in the correct pipelines in the language repo. Usually the core or the template pipeline but really any pipeline that will trigger based on eng/common changes.

File: eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+jobs:
+  - job: EnforceCommonWorkflow
+
+    pool:
+      vmImage: windows-2019
+
+    steps:
+      - pwsh: |
+          $PrSourceBranch = $(System.PullRequest.SourceBranch)
```
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml:19]: If this isn't a PR this step will fail in a weird way so I suggest conditioning it on the whether or not is a PR.

File: eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+          Write-Host "[debug]Pull Request Source Branch: ${PrSourceBranch}"
+          if (-not $PrSourceBranch.StartsWith("sync-eng/common"))
```
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml:21]: You should be able to produce a diff of the files changed and if there is anything under eng/common then enable this check.

File: eng/common/scripts/Invoke-GitHubAPI.ps1 (hunk 2 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Invoke-GitHubAPI.ps1:108]: Was going to use it then figured I did not need to. I still think its a good feature though.

File: eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml (hunk 3 above)
Comment (chidozieononiwu) [reply eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml:3]: My idea was to use the pipeline pr trigger functionality, so the job should not even run if there is not change to eng/common

File: eng/common/scripts/Invoke-GitHubAPI.ps1 (hunk 2 above)
Comment (weshaggard) [reply eng/common/scripts/Invoke-GitHubAPI.ps1:108]: Lets remove it until we have some that needs to use it.

File: eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml (hunk 3 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/jobs/eng-common-workflow-enforcer.yml:3]: I'm not convinced that having a dedicated pipeline is worth it for this check.

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+steps:
+  - ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+    - pwsh: |
+        $PrSHA = "$(Build.SourceVersion)"
+        $masterSHA = git rev-parse origin/master
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:5]: We should not assume master. We should use the source and target branch. See https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/get-markdown-files-from-changed-files.ps1#L5

I think you can do something as simple as:
```
git diff "origin/${env:SYSTEM_PULLREQUEST_TARGETBRANCH}" HEAD --name-only -- 'eng/common/*'
```

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        {
+          if ($line -match "eng/common")
+          {
+            $filesInCommonDir += $line
+          }
+        }
+
+        if ((-not $PrSourceBranch.StartsWith("sync-eng/common")) -and ($filesInCommonDir.Count -gt 0))
+        {
+          Write-Host "##vso[task.LogIssue type=warning;]Changes to the files below (files in eng/common directory) should not be made on this Repo`n${filesInCommonDir}"
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:19]:
```suggestion
          Write-Host "##vso[task.LogIssue type=warning;]Changes to files under 'eng/common` directory should not be made in this Repo`n${filesInCommonDir}"
```

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+          Write-Host "##vso[task.LogIssue type=warning;]Please follow workflow at https://github.com/Azure/azure-sdk-tools/blob/master/doc/common/common_engsys.md"
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:20]: Should these be errors instead of warnings? Especially given that this step will fail.

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+          exit 1
+        }
+      displayName: Prevent eng/common Changes from non Sync Prs
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:23]: Something to keep in mind this step will cause the CI legs to fail anytime we we are trying to test changes in the language repo's before we merge them into the language repo. We might want to consider some way to opt-out of this beyond just the source branch name.

Comment (chidozieononiwu) [reply eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml]: I added a condition to the step.

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    - pwsh: |
+        $filesInCommonDir = git diff "origin/$(System.PullRequest.TargetBranch)" HEAD --name-only -- 'eng/common/*'
+
+        if ((-not $PrSourceBranch.StartsWith("sync-eng/common")) -and ($filesInCommonDir.Count -gt 0))
+        {
+          Write-Host "##vso[task.LogIssue type=error;]Changes to files under 'eng/common' directory should not be made in this Repo`n${filesInCommonDir}"
+          Write-Host "##vso[task.LogIssue type=error;]Please follow workflow at https://github.com/Azure/azure-sdk-tools/blob/master/doc/common/common_engsys.md"
+          exit 1
+        }
+      displayName: Prevent eng/common Changes from non Sync Prs
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:12]:
```suggestion
      displayName: Prevent changes to eng/common outside of azure-sdk-tools repo
```

File: eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+      condition: and(succeeded(), ne(variables['Skip.EngCommonWflEnforcer'], 'true'))
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml:13]:
```suggestion
      condition: and(succeeded(), ne(variables['Skip.EngCommonWorkflowEnforcer'], 'true'))
```

**Review by weshaggard — APPROVED** [review]
A couple minor tweaks but otherwise looks good.

**Review by weshaggard — APPROVED** [review]
